src/cgroups/v2/systemd_manager.rs

```diff
@@ -0,0 +1,304 @@
+            .into_iter()
+            .map(|c| format!("{}{}", "+", c.to_string()))
+            .collect();
+
+        // Write the controllers to the root_path.
+        Self::write_controllers(&self.root_path, &controllers)?;
+
+        let mut current_path = self.root_path.clone();
+        let mut components = cgroups_path.components().skip(1).peekable();
+        // Verify that *each level* in the downward path from the root cgroup
+        // down to the cgroup_path provided by the user is a valid cgroup hierarchy.
+        // containing the attached controllers.
+        while let Some(component) = components.next() {
+            current_path = current_path.join(component);
+            if !current_path.exists() {
+                fs::create_dir(&current_path)?;
+                fs::metadata(&current_path)?.permissions().set_mode(0o755);
+            }
+
+            // last component cannot have subtree_control enabled due to internal process constraint
+            // if this were set, writing to the cgroups.procs file will fail with Erno 16 (device or resource busy)
+            if components.peek().is_some() {
+                Self::write_controllers(&current_path, &controllers)?;
+            }
+        }
+
+        write_cgroup_file(full_path.join(CGROUP_PROCS), &pid.to_string())?;
+        Ok(full_path)
+    }
+
+    fn get_available_controllers<P: AsRef<Path>>(
+        &self,
+        cgroups_path: P,
+    ) -> Result<Vec<ControllerType>> {
+        let controllers_path = self.root_path.join(cgroups_path).join(CGROUP_CONTROLLERS);
+        if !controllers_path.exists() {
+            bail!(
+                "cannot get available controllers. {:?} does not exist",
+                controllers_path
+            )
+        }
+
+        let mut controllers = Vec::new();
+        for controller in fs::read_to_string(&controllers_path)?.split_whitespace() {
+            match controller {
+                "cpu" => controllers.push(ControllerType::Cpu),
+                "io" => controllers.push(ControllerType::Io),
+                "memory" => controllers.push(ControllerType::Memory),
+                "pids" => controllers.push(ControllerType::Pids),
+                _ => continue,
+            }
+        }
+
+        Ok(controllers)
+    }
+
+    fn write_controllers(path: &Path, controllers: &Vec<String>) -> Result<()> {
+        for controller in controllers {
+            common::write_cgroup_file_str(path.join(CGROUP_SUBTREE_CONTROL), controller)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl CgroupManager for SystemDCGroupManager {
+    fn apply(&self, linux_resources: &LinuxResources, pid: Pid) -> Result<()> {
+        // Dont attach any pid to the cgroup if -1 is specified as a pid
+        if pid.as_raw() == -1 {
+            return Ok(());
+        }
+        let full_cgroup_path = self.create_unified_cgroup(pid)?;
+
+        for controller in CONTROLLER_TYPES {
+            match controller {
```

Comment:
I assume this is temporary until you tell systemd to do this for you?

Reply:
No. this still has to be done here (I might be mistaken though)... the only thing that would be different is adding the unit (via dbus) to systemd - that way systemd manages the container and can do different things with it (like start the container on boot etc.). I might be mistaken though. I will try and look at what `runc` does.

Reply:
Hmm looking at the `runc` [docs](https://github.com/opencontainers/runc/blob/6023d635d725a74c6eaa11ab7f3c870c073badd2/docs/systemd.md#resource-limits) it might be the case that the above resource limitations wont be needed and indeed need to be set as properties on the unit created.

Reply:
See also https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html. I would recommend to split this into controllers again, that then can be done by other people like we did with cgroup v1 and v2.

Reply:
Sounds good. Thanks!